### PR TITLE
Upgrade to wgpu 23

### PR DESCRIPTION
**Pull request by tychedelia**
Fixes https://github.com/bevyengine/bevy/issues/15893

### CODE REVIEW

File: crates/bevy_color/Cargo.toml [hunk 1]
```diff
@@ -23,7 +23,7 @@ derive_more = { version = "1", default-features = false, features = [
   "from",
   "display",
 ] }
-wgpu-types = { version = "22", default-features = false, optional = true }
+wgpu-types = { git = "https://github.com/gfx-rs/wgpu", branch = "trunk", default-features = false, optional = true }
```
Comment (djeedai) [inline crates/bevy_color/Cargo.toml:26]: It seems dangerous to move to a trunk branch no? Shouldn't we specify a locked commit hash instead? (even if it's the same as v23 at the time of writing)

**Review by djeedai — COMMENTED** [review]
Looks good to merge as long as we're OK to take a dependency on a moving target (I don't think we should). Also maybe the danger of upgrading to a new version close to release (but I understand it fixes some bug).

File: crates/bevy_color/Cargo.toml (hunk 1 above)
Comment (rparrett) [reply crates/bevy_color/Cargo.toml:26]: wgpu 0.23 isn't released yet, presumably the author intends to update this and take the PR out of draft status once it is released.

**Comment by JMS55** [comment]
No need to wrap bind groups in Some() anymore https://github.com/gfx-rs/wgpu/pull/6452

**Comment by JMS55** [comment]
Actually that makes it worse... Can't automatically convert between bevy's and wgpu's bind group type...
```
note: required by a bound in `wgpu::RenderPass::<'encoder>::set_bind_group`
   --> C:\Users\Jasmine\.cargo\git\checkouts\wgpu-ebe3fb3e55696655\a9033f5\wgpu\src\api\render_pass.rs:80:26
    |
77  |     pub fn set_bind_group<'a>(
    |            -------------- required by a bound in this associated function
...
80  |         bind_group: impl Into<Option<&'a BindGroup>>,
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RenderPass::<'encoder>::set_bind_group`

error[E0277]: the trait bound `std::option::Option<&wgpu::BindGroup>: From<&bind_group::BindGroup>` is not satisfied
   --> crates\bevy_render\src\view\window\screenshot.rs:626:36
    |
626 |             pass.set_bind_group(0, &prepared_state.bind_group, &[]);
    |                  --------------    ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<&bind_group::BindGroup>` is not implemented for `std::option::Option<&wgpu::BindGroup>`, which is required by `&bind_group::BindGroup: Into<std::option::Option<&wgpu::BindGroup>>`
    ```

**Comment by AlexAegis** [comment]
https://github.com/gfx-rs/wgpu/pull/6465

Looks like wgpu 23 is about to release! 🎉

**Comment by Elabajaba** [comment]
This doesn't build on iOS currently according to Francois' testing.

**Comment by mockersf** [comment]
Just re-triggered a build, here is the failure: https://github.com/TheBevyFlock/bevy-example-runner/actions/runs/11533949819/job/32107445980

**Comment by tychedelia** [comment]
> Just re-triggered a build, here is the failure: https://github.com/TheBevyFlock/bevy-example-runner/actions/runs/11533949819/job/32107445980

From last month:

https://github.com/gfx-rs/wgpu/commit/fb0cb1eb11663f8de023f6dd64128ed1f5342ec7#diff-46fec780a1b7071adf9590bb9d55dfea0f97bd15419ccbe55ee89607fc244335R27

**Comment by BenjaminBrienen** [comment]
With `cargo run --example relative_cursor` I have to set the `web-sys` dependency to `0.3.72` and then it compiles.

**Comment by mockersf** [comment]
> Edit: pushed a commit to attempt to fix this.

Building now works 🎉

**Comment by JMS55** [comment]
Might be worth letting wgpu people know that they should include it in the changelog for wgpu 23.

**Comment by mockersf** [comment]
released! https://github.com/gfx-rs/wgpu/releases/tag/v23.0.0

**Comment by JMS55** [comment]
Something we can finally now do is use const arrays with dynamic indexes, for e.g. shadow sampling patterns. 

We should take advantage of this to cleanup a bunch of code.

**Comment by djeedai** [comment]
FYI this also brings CAS atomic operations to Metal. I didn't check if that was the last platform missing them (I expect wasm needs CORS, as with everything atomic due to security reasons, if it even supports CAS at the API level already).

**Comment by mockersf** [comment]
@tychedelia naga_oil 0.16 has been updated and released

**Review by kristoff3r — APPROVED** [review]
LGTM, I also tested a bunch of 2d/3d examples on Linux and WebGL and stuff seems to work.

**Review by mockersf — APPROVED** [review]

**Comment by mockersf** [comment]
I would like to merge this one quickly and publish a new rc soon after, so get your reviews now! well in the next day or so

**Review by JMS55 — APPROVED** [review]

**Pull request merged**